pression/data/Compressor.h

```diff
@@ -0,0 +1,126 @@
+#include <lunchbox/debug.h> // LBUNIMPLEMENTED
+
+namespace pression
+{
+namespace data
+{
+/** Base interface for lossless CPU compressors of binary data. */
+class Compressor
+{
+public:
+    virtual ~Compressor() {} //!< @internal
+
+    typedef lunchbox::Bufferb Result; //!< Single result data buffer
+    typedef std::vector< Result > Results; //!< Set of result chunks
+
+    /**
+     * Compress the given data and return the result.
+     *
+     * This default implementation will use getCompressBound(), getChunkSize()
+     * and call the protected compress() method to parallelize the
+     * compression. The result is valid until the next call to compress or
+     * destruction of this instance.
+     *
+     * @param data pointer to data to compress
+     * @param size number of bytes to compress
+     * @return the compressed data chunk(s)
+     */
+    PRESSIONDATA_API virtual const Results& compress( const uint8_t* data,
+                                                      size_t size );
+    /**
+     * Decompress the given data.
+     *
+     * This default implementation will decompress the given input in parallel,
+     * assuming getChunkSize() returns the same value as during the compress()
+     * operation.
+     *
+     * @param input compressed data chunk(s) produced by compress()
+     * @param sizes compressed data chunk size(s) produced by compress()
+     * @param data pointer to pre-allocated memory for the decompressed data
+     * @param size decompressed data size
+     * @throw std::runtime_error if chunksize does not match input
+     */
+    PRESSIONDATA_API virtual
+    void decompress( const std::vector< const uint8_t* >& input,
+                     const std::vector< size_t >& sizes,
+                     uint8_t* data, size_t size );
+
+    /** @overload convenience wrapper */
+    PRESSIONDATA_API void decompress( const Results& input, uint8_t* data,
+                                      size_t size );
+
+    /** @return the result of the last compress() operation. */
+    const Results& getCompressedData() const { return compressed; }
+
+protected:
+    Compressor() {}
+    Compressor( const Compressor& ) = delete;
+    Compressor( Compressor&& ) = delete;
+    Compressor& operator = ( const Compressor& ) = delete;
+    Compressor& operator = ( Compressor&& ) = delete;
+
+    /** @return an upper bound of the compressed output for a given size. */
+    virtual size_t getCompressBound( const size_t size ) const = 0;
+
+    /** @return the optimal chunk size for this compressor */
+    virtual size_t getChunkSize() const { return LB_64KB; }
+
+    /**
+     * Compress the given chunk.
+     *
+     * @param data pointer to data to compress
+     * @param size number of bytes to compress
+     * @param output pre-allocated output chunk
+     */
+    virtual void compress( const uint8_t* data LB_UNUSED, size_t size LB_UNUSED,
+                           Result& output LB_UNUSED ) { LBUNIMPLEMENTED }
```

Comment:
My gut tells me it's better to rename this method (and the other one below) to compressImpl to avoid confusion with the public API, as they are not easy to differentiate by the propotype at first glance.